### ISSUE TITLE
[ibex] Shadow reset for shadow core

### ISSUE THREAD

**Issue by tjaychen** [lowRISC/opentitan#8538]
Should we consider wiring a shadow reset to second lock step core? 
It might not matter I think... if the reset net is hit, the two core results won't agree anyways, thus triggering an alert,

But I wanted to ask the question and see if everyone agrees.

**Comment by tjaychen** [comment]
@cdgori 
@msfschaffner 
@GregAC

**Comment by GregAC** [comment]
We've already got a delayed reset generation for the lockstep core, which resets it in lock step the appropriate number of cycles after the main core (https://github.com/lowRISC/ibex/blob/97fa5cf28024b882ffa1acbe81f6b6c865715840/rtl/ibex_lockstep.sv#L109-L159) does this achieve what you were looking for @tjaychen ?

**Comment by tjaychen** [comment]
not completely. If i'm reading this right, if the normal `rst_ni` were glitched, both the normal core reset and the shadow reset would trigger right? The idea of that shadow reset is that unless you hit an upstream node, one reset should not impact the other.

However, this idea does not apply quite as cleanly to shadow core since I think you guys have exact cycle requirements. I would say let's table this request to future enhancement and leave things as is for now.

**Comment by johannheyszl** [comment]
@GregAC I would suggest to close as not planned since the lock step configuration would generate alerts on mismatch if only one core resets through fault.

**Comment by GregAC** [comment]
@johannheyszl  what can happen is if a glitch occurs on the (single) reset that goes from the rstmgr to Ibex this could succeed in generating a false reset without an alert if it glitches before we do the delayed reset generation for the shadow core.

Note this same attack would apply to all the blocks in the design though it may be more critical against Ibex as it's the CPU.

**Comment by johannheyszl** [comment]
Thanks @GregAC : in case of a glitch to the primary core where the shadow core reset logic misses it and does not reset the shadow core, the two cores would diverge and the lockstep config would detect this 'very soon' though, correct?

**Comment by GregAC** [comment]
Yes exactly

**Comment by johannheyszl** [comment]
Thx @GregAC 
@vogelpi I suggest we close as not planned because it seems like a smaller defense in depth addition, until any findings would suggest otherwise.

**Comment by vogelpi** [comment]
TL;DR: It's fine to close this as not planned.

To be on the safe side @johannheyszl, @nasahlpa  and I had another look at this together. The idea of adding a separate shadow reset for the shadow core that is driven by a separate node in the reset manager (similarly to what is done for shadowed CSRs inside peripherals) is to be certain that it's impossible to reset both main and shadow core partially in way not detectable by lockstep. One could imagine and adversary e.g. trying to disable interrupts by glitching the reset line of the corresponding `mie` CSRs (these are in the core itself).

What we've found is that the reset of the shadow core `rst_shadow_n` is driven by a clock mux and the relevant input to this clock mux is coming out of an FF. This FF (`rst_shadow_set_q`) is using the same reset `rst_ni` as the main core. See https://github.com/lowRISC/ibex/blob/b6d8b9f07582cd9f77b5c1f1680ad1ab68532796/rtl/ibex_lockstep.sv#L145-L164 
This means:
- A glitch on the shadow reset `rst_shadow_n` will always be detected at some point by the lockstep because it only affects the shadow core.
- A glitch on the main reset `rst_ni` either
  - Resets something inside the main core only which is going to be detected eventually by the lockstep, or
  - Resets something inside the main core and/or the flop `rst_shadow_set_q` driving the shadow reset `rst_shadow_n`. If the FF is not hit, lockstep will detect the glitch eventually. If the FF is hit, the reset of the shadow core will be active until the next active clock edge. Depending on how long it takes for the active clock edge to come, either
     - most of the shadow core is going to be reset which will get detected by the lockstep, or
     - the glitch will propagate to just a few FFs, i.e., we have glitch on the shadow and the main core's resets which is not ideal.

However, due to the the different boundary conditions of both cores (main core connects to TL-UL, RV_DM, PLIC etc. whereas the shadow core only interfaces to FFs, i.e., the timing is for example much more relaxed for the latter) it's very unlikely that a glitch on the main reset is going to hit the same FFs inside the main and shadow core without detection by the lockstep. Exploiting this in a useful way is highly unlikely. Due to all of this, adversaries will most likely explore different attack paths. Thus, there seems little to gain here compared to the effort required for adding the separate reset node/tree.
